secret-mounter: codebase review prior to merge

shlokchaudhari9 opened the pull request:
This PR consists of the e2e solution for mentioned problem statement in the README.md

shlokchaudhari9 commented:
@viveksinghggits @PrasadG193 Please review the PR

viveksinghggits commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN mkdir -p bin
+RUN go build -o bin/
+
+ENTRYPOINT ["/go/src/secret-sneaker/bin/secret-sneaker"]
```
add a new line at the end of the file.

viveksinghggits commented on README.md:
```diff
@@ -10,7 +10,15 @@ We want to enable users to mount secrets on a workload (deployment/statefulset)
 Once the application is running, user can "sneak-in" (create) secret as a VolumeMount automatically along with a deployment creation.
 
-Just update your secrets as a key:value pair at path /mnt/data/secret-sneaker-data/secrets.json on your host machine (Create the entire path if required)
+Just update your secrets as a key:value pair at path /mnt/data/secret-sneaker-data/secrets.json on your host machine (Create the entire path if required) as shown in sample JSON below;
```
1. why do I have to create the json?
2. Why at the specific location?

also 
```
Sneak secret file in your namespace automatically while creating a deployment
```
this didn't make sense to me.

viveksinghggits commented on README.md:
````diff
@@ -38,22 +46,24 @@ spec:
         name: busybox
 ```
-Resultant deployment will have secrets present under path /etc/secret-sneaker-data/
+Resultant deployment will have secrets present under path /etc/secret-sneaker-data/ inside the pod container
````
can this path be taken from user?

viveksinghggits replied:
also what do I have to do, if I have to just mount one key value pair from the secret?

shlokc9 replied:
> can this path be taken from user?

Hey @viveksinghggits I am not sure how we can get the path dynamically from the user without a templating tool like Ansible or Terraform. I tried to use the Labels/Annotations key:values for the path as well but Valid value regex doesn't allow "/" character in the string.

viveksinghggits commented on src/custom/controller.go:
```diff
@@ -0,0 +1,177 @@
+	return true
+}
+
+func (cntrlreceiver *controller) createSecret(ns, name string) error {
```
why are we creating the secret, I don't remember specifying that user will not create the secret.

viveksinghggits commented on src/custom/controller.go:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	if val, ok := deployment.Labels["app"]; ok && val == "secret-sneaker" {
+
```
we shouldn't have un necessary new lines.

viveksinghggits commented on src/custom/controller.go:
```diff
@@ -0,0 +1,177 @@
+		fmt.Println("Error in cntrlreceiver.depLister.Deployments(ns).Get(): ", err.Error())
+	}
+
+	if val, ok := deployment.Labels["app"]; ok && val == "secret-sneaker" {
```
`secret-snekaer` should be declared as const.

viveksinghggits commented on src/custom/controller.go:
```diff
@@ -0,0 +1,177 @@
+		},
+	}
+	deployment.Spec.Template.Spec.Volumes = append(deployment.Spec.Template.Spec.Volumes, secretvolume)
+	containervolumemount := corev1.VolumeMount{
```
we use camel cases, I see you have not used them anywhere.

viveksinghggits replied:
we -> Go

shlokchaudhari9 commented:
Hey @viveksinghggits I have made the changes as per our discussions in the morning sync-up call. Please check the README.md file as well for any issues with the flow. https://github.com/shlokchaudhari9/secret-mounter/tree/stable-application

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+// Optional field
+const DeploymentLabelSecretKeys string = "secret-keys"
+// Secret Keys separator
+const SecretKeysSeparator string = "."
```
consts can be grouped 
```suggestion
const (
    SecretMounter string = "secret-mounter"
    DefaultContainerSecretPath string = "/etc/" + SecretMounter + "-data/"

    // Mandatory field
    DeploymentLabelSecretName string = "secret-name"
    // Optional field
    DeploymentLabelSecretKeys string = "secret-keys"
    // Secret Keys separator
    SecretKeysSeparator string = "."
)
```

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+
+func (cntrlReceiver *controller) worker() {
+	for cntrlReceiver.processItem() {
+	}
```
can we make the function more readable? maybe Get the item here a and pass it as a arg to the function

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+	defer cntrlReceiver.workQueue.Forget(item)
+	key, err := cache.MetaNamespaceKeyFunc(item)
+	if err != nil {
+		fmt.Println("Error in cache.MetaNamespaceKeyFunc(): ", err.Error())
```
return and handle error?

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+	ctx := context.Background()
+	deployment, err := cntrlReceiver.depLister.Deployments(ns).Get(name)
+	if err != nil {
+		fmt.Println("Error in cntrlReceiver.depLister.Deployments(ns).Get(): ", err.Error())
```
please return and handle error

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+func (cntrlReceiver *controller) Run(stopCh <-chan struct{}) {
+	fmt.Println("Starting Custom Controller....")
+	if !cache.WaitForCacheSync(stopCh, cntrlReceiver.depCacheSynced) {
+		fmt.Println("Waiting for the cache to be synced....")
```
Instead of Print statements for info and error messages, please use loggers like - https://pkg.go.dev/github.com/sirupsen/logrus or https://pkg.go.dev/k8s.io/klog

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+				fmt.Println("Error in cntrlReceiver.mountSecretInDep(): ", err.Error())
+				return err
+			}
+		} else {
+			fmt.Println("Secret name not found in the deployment metadata.labels - Skipping secret mount....")
+		}
```
else condition is not needed if we add return at the end of if statements
```suggestion
                        return nil
		}
		fmt.Println("Secret name not found in the deployment metadata.labels - Skipping secret mount....")

```

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+	ctx context.Context, ns, name string, secret coreV1.Secret, depLabels map[string]string) error {
+	deployment, err := cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(ctx, name, metaV1.GetOptions{})
+	if err != nil {
+		fmt.Println("Error in cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(): ", err.Error())
```
return and handle error

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		fmt.Println("Error in cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(): ", err.Error())
+	}
+	secretVolume := coreV1.Volume{
+		Name: secret.Name + "-secret-volume",
+		VolumeSource: coreV1.VolumeSource{
+			Secret: &coreV1.SecretVolumeSource{
+				SecretName: secret.Name,
+			},
+		},
+	}
+	containerVolumeMount := coreV1.VolumeMount{
+		Name:      secret.Name + "-secret-volume",
+		MountPath: DefaultContainerSecretPath,
+		ReadOnly:  true,
+	}
```
This can be moved to a helper function e.g something like getDeployVolumeMount()

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+func (cntrlReceiver *controller) mountSecretInDep(
+	// Fetching the latest version of deployment and modifying it
+	ctx context.Context, ns, name string, secret coreV1.Secret, depLabels map[string]string) error {
+	deployment, err := cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(ctx, name, metaV1.GetOptions{})
```
the comment should be either on top of the func name or inside func def

```suggestion
// mountSecretInDep Fetching the latest version of deployment and modifying it
func (cntrlReceiver *controller) mountSecretInDep(
	ctx context.Context, 
	ns string,
	name string,
	secret coreV1.Secret,
	depLabels map[string]string) error {
	deployment, err := cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(ctx, name, metaV1.GetOptions{})
```

PrasadG193 commented on src/custom/controller.go:
```diff
@@ -0,0 +1,175 @@
+		workQueue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), SecretMounter),
+	}
+	depInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: newController.handleAdd,
+		},
```
@shlokchaudhari9 what will happen if we add labels to existing deployment? We are not handling update events?

shlokc9 replied:
I have updated the code to handle existing deployment updates as well

PrasadG193 commented on test/test.yaml:
```diff
@@ -0,0 +1,38 @@
+kind: Deployment
+metadata:
+  labels:
+    app: test-deployment
+    # Mandatory field
+    app: secret-mounter
```
@shlokchaudhari9 why do we need this label `app`?

shlokc9 replied:
Right now app=secret-mounter & secret-name=[secret-name] labels both are checked to start mounting secrets on the deployment. But, @viveksinghggits pointed out that we can just drop the app=secret-mounter label and only use secret-name=[secret-name] as primary check to start mounting. I'll be doing this update as well.

shlokchaudhari9 commented:
@PrasadG193 I have addressed all your comments except replacing print statements with loggers. Also ran golangci-lint on the code and made appropriate lint fixes as well.

shlokchaudhari9 commented:
@PrasadG193 Updated code base with all your suggested comments. PTAL
